ibacm/src/acm.c

```diff
@@ -2070,7 +2070,9 @@ static void acm_ep_ip_iter_cb(char *ifname, union ibv_gid *gid, uint16_t pkey,
 
 	dev = acm_get_device_from_gid(gid, &port_num);
 	if (dev && ep->port->dev == dev
-	    && ep->port->port.port_num == port_num && ep->endpoint.pkey == pkey) {
+	    && ep->port->port.port_num == port_num &&
+		/* pkey retrieved from ipoib has always full mmbr bit set */
+		(ep->endpoint.pkey | 0x8000) == pkey) {
```

Comment:
I'm a bit worried about this.  Unless the user is using AF_IB I don't believe the Pkey will be set correctly to use the AF_INET/INET6 address assigned to the IPoIB interface.  I think the error should stand.  OR a Pkey table search should be used in the case of an EP being addressed with AF_IB.

Reply:
If you echo a limited pkey into /sys/class/net/ipoib/create_child, the new device's /sys/class/net/ib0.0123/pkey will be the full partition. Isn't that correct?

Reply:
Yes but IPoIB will not be able to use that partition if the membership of the port is limited for that partition.

Reply:
@weiny2 Not quite right. If the peer is a full member, communication flows smoothly. This is actually how orcl multitenant virtualized rdbms works, which is the use-case requiring this bug-fix.

Reply:
How can you have more than 1 limited member port in a subnet?  Mcast traffic would fail between any 2 limited members.  I guess if there is only 1 limited member you are ok but that seems to be a pretty narrow use case.

Reply:
@weiny2 You are right that limited<->limited cannot communicate, and that is why we use limited partitions :-) VMs from different tenants shall not be able to communicate, and they do not. But they need I/O, so the array of I/O servers are full members of this partition. Trust me, orcl has actively been using partition in the IB fabric for a while.

Reply:
No, the IPoIB subnet shall **not** be able to ping between node A and B (since they are different tenants in my case). That is the virtue of limited partitions. But A and B can communicate with X. And that is my use case.

There is no Spoofing. The fact is that if IPoIB runs on a limited partition, its sysfs returns the full member, which at least in this context, is wrong. One could argue to fix IPoIB in this context, I can agree to that.

Try for yourself, make a limited partition and make ibacm work with it. It doesn't!

Are you saying that a large number of Exadata employments using VMs and partitioning is deemed not to work if you toss ibacm into the equation? Well, they do not right now, but this commit fixes that, and I do not see what harm it causes.

As stated, the IPoIB sysfs returns always the full membership partition. If the actual membership is full, ORing in 0x8000 is a NOP. If the actual membership is limited, the ORing makes a match for the endpoint, and makes ibacm useful.

Reply:
@weiny2 Did what I wrote above make any sense to you?

Reply:
My apologies.  I see that the comment I meant to submit a few days ago was marked "Pending".  I believe it is submitted now.